library/src/main/java/j2html/rendering/IndentedHtml.java

```diff
@@ -0,0 +1,247 @@
+     * the given Config.
+     *
+     * @param config The Config which will specify indentation, text escapement, tag closing, etc.
+     * @param <T>    The type of the Appendable to which HTML will be appended.
+     * @return An HtmlBuilder for indented HTML.
+     */
+    public static final <T extends Appendable> IndentedHtml inMemory(Config config) {
+        return into(new StringBuilder(), config);
+    }
+
+    private final T out;
+    private final Indenter indenter;
+    private final TextEscaper textEscaper;
+    private final TagBuilder enclosingElementAttributes;
+    private final TagBuilder emptyElementAttributes;
+
+    // Dealing with preformatted elements (pre and textarea) requires
+    // that we know what our parent elements are.  To do that we use
+    // a stack; adding items as start tags are created, and removing them
+    // as those tags are closed.  Determining whether or not we are
+    // currently rendering into a preformatted element is as simple as
+    // asking if any tags on the stack match a preformatted element name.
+    private Stack<String> trace = new Stack<>();
+
+    private IndentedHtml(T out, Config config) {
+        this.out = out;
+        this.indenter = config.indenter();
+        this.textEscaper = config.textEscaper();
+        this.enclosingElementAttributes = new IndentedTagBuilder(false);
+        this.emptyElementAttributes = new IndentedTagBuilder(config.closeEmptyTags());
+    }
+
+    private boolean isContentSelfFormatting() {
+        return trace.contains("pre") || trace.contains("textarea");
+    }
+
+    private int lvl() {
+        return trace.size();
+    }
+
+    @Override
+    public TagBuilder appendStartTag(String name) throws IOException {
+        if (!isContentSelfFormatting()) {
+            out.append(indenter.indent(lvl(), ""));
+        }
+
+        trace.push(name);
+
+        out.append("<").append(name);
+        return enclosingElementAttributes;
+    }
+
+    @Override
+    public HtmlBuilder appendEndTag(String name) throws IOException {
+        if (!name.equals(trace.peek())) {
+            throw new RuntimeException("Incorrect element closed: " + name + ".  Expected: " + trace.peek());
+        }
+
+        if (!isContentSelfFormatting()) {
+            trace.pop();
+            out.append(indenter.indent(lvl(), ""));
+        } else {
+            trace.pop();
+        }
+
+        out.append("</").append(name).append(">");
+
+        if (!isContentSelfFormatting()) {
+            out.append("\n");
+        }
+
+        return this;
+    }
+
+    @Override
+    public TagBuilder appendEmptyTag(String name) throws IOException {
+        if (!isContentSelfFormatting()) {
+            out.append(indenter.indent(lvl(), ""));
+        }
+        out.append("<").append(name);
+        return emptyElementAttributes;
+    }
+
+    private void appendLines(String txt) throws IOException {
+        if (!isContentSelfFormatting()) {
```

Comment:
```suggestion
        if (!isContentSelfFormatting() && txt != null) {
```
Background: I got a `NullPointerException` because I had a `text(null)` somewhere in my code.
Alternatively we might prohibit `null` already when constructing the HTML tree, but I guess that's a bigger change.

Reply:
I just thought about it - to keep the indent right the better solution is
`            String[] lines = String.valueOf(txt).split("\n");`
in the next line (and probably there should also be a test for `null` values)

Reply:
That can handle the situation where txt is being indented, but would still risk a NPE within _pre_ an _textarea_ elements.  Perhaps we can handle this at a higher level with `String.valueOf()` in the `Text` and `UnescapedText` classes.  That way the builders can avoid null checks on everything, and we can begin limiting the spread of nulls.